src/witan/gateway/handler.clj

```diff
@@ -1,49 +1,131 @@
+  (let [edn (pr-str m)]
+    (log/debug "Sending EDN:" edn)
+    (send! ch edn)))
+
+(defn send-message!
+  [ch m]
+  (if-let [err (rs/check-message "1.0" m)]
+    (log/error "Failed to send a message - validation failed:" err)
+    (send-edn! ch m)))
+
+(defn dispatch-event!
+  [key version params receipt]
+  (when-let [{:keys [ch]} (get @receipts receipt)]
+    (try
+      (send-message! ch {:message/type :event
+                         :event/key key
+                         :event/version version
+                         :event/params params
+                         :event/created-at (iso-dt-now)
+                         :command/receipt receipt})
+      (catch Exception e
+        (println "Failed to dispatch event:" e)))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Message Handling
+
+(defmulti handle-message
+  (fn [ch msg components] (:message/type msg)))
+
+(defmethod handle-message
+  :query
+  [ch {:keys [query/edn query/id]} {:keys [queries]}]
+  (if-not (vector? edn)
+    (send-message! ch {:message/type :query-response :query/id id :query/error "Query needs to be a vector"})
+    (let [results (mapv (partial p/route-query queries) edn)]
+      (send-message! ch {:message/type :query-response :query/id id :query/results results}))))
+
+;;
+
+(defmethod handle-message
+  :command
+  [ch {:keys [command/id command/key command/version] :as command} {:keys [kafka]}]
+  (let [receipt (java.util.UUID/randomUUID)
+        now (iso-dt-now)]
+    (swap! receipts assoc receipt {:ch ch :at (t/now)})
+    (if-let [error (command/receive-command!
+                    receipt
+                    (assoc command :command/created-at now)
+                    kafka)]
+      (do
+        (log/error "Failed to send msg to Kafka:" error)) ;; TODO do we tell the client?
+      (send-message! ch {:message/type :command-receipt
+                         :command/key key
+                         :command/version version
+                         :command/id id
+                         :command/receipt receipt
+                         :command/received-at now}))))
+
+(defmethod handle-message
+  :default
+  [_ msg _]
+  (println "UNKNOWN MESSAGE:" (:type msg) msg))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Connection Handling
+
+(defn disconnect! [channel status]
+  (swap! channels #(remove #{channel} %)))
+
+(defn connect! [channel]
+  (swap! channels conj channel))
+
+(defn ws-handler [request]
+  (let [components (:witan.gateway.components.server/components request)]
+    (with-channel request channel
+      (connect! channel)
+      (on-close channel (partial disconnect! channel))
+      (on-receive channel #(try
+                             (let [msg (read-string %)
+                                   error (rs/check-message "1.0" msg)]
+                               (if-not error
+                                 (handle-message channel msg components)
```

Comment:
No, just a single channel. The `handle-message` `defmulti` dispatches on whether it's a query or a command.